racket/collects/racket/stream.rkt

```diff
@@ -179,3 +184,129 @@
+
+(define (chaperone-stream s first-proc rest-proc . props)
+  (chaperone-generics
+   gen:stream
+   ; the struct-info param is a hack, see PR 14970
+   (apply chaperone-struct s struct-info (λ (a b) (values a b)) props)
+   [stream-first
+    (λ (stream-first)
+      (chaperone-procedure stream-first
+                           (λ (s) (values (λ (v) (first-proc v)) s))))]
+   [stream-rest
+    (λ (stream-rest)
+      (chaperone-procedure stream-rest
+                           (λ (s) (values (λ (v) (rest-proc v)) s))))]))
+
+;; Stream contracts ----------------------------------------------------------------------------------
+
+(define (stream/c-name ctc)
+  (define elem-name (contract-name (base-stream/c-content ctc)))
+  (apply build-compound-type-name
+         'stream/c
+         elem-name
+         '()))
+
+(define (add-stream-context blame)
+  (blame-add-context blame "a value generated by"))
+
+(define (check-stream/c ctc val blame)
+  (unless (stream? val)
+    (raise-blame-error blame val '(expected "a stream" given: "~e") val)))
+
+(define (check-stream/c-np ctc val blame)
+  (if (stream? val)
+      #f
+      (λ (neg-party)
+        (raise-blame-error blame #:missing-party neg-party
+                           val '(expected "a stream" given: "~e") val))))
+
+(define (stream/c-stronger? a b)
+  (contract-stronger? (base-stream/c-content a) (base-stream/c-content b)))
+
+; streams are lazy, so we need to contract the rest of the stream lazily (which can be a list)
+(define (contract-stream-rest v ctc blame)
+  (define elem-ctc (base-stream/c-content ctc))
+  (define new-ctc (if (list? v) (listof elem-ctc) ctc))
+  (contract new-ctc v
+            (blame-positive blame) (blame-negative blame)
+            (blame-value blame) (blame-source blame)))
+
+(define ((ho-val-first-projection impersonate/chaperone-stream) ctc)
+  (define elem-ctc (base-stream/c-content ctc))
+  (define vfp (get/build-val-first-projection elem-ctc))
+  (λ (blame)
+    (define stream-blame (add-stream-context blame))
+    (define pos-elem-proj (vfp stream-blame))
+    (define neg-elem-proj (vfp (blame-swap stream-blame)))
+    (λ (val)
+      (or (check-stream/c-np ctc val blame)
+          (λ (neg-party)
+            (if (list? val)
+                (contract-stream-rest val ctc blame)
+                (impersonate/chaperone-stream 
+                 val
+                 (λ (v) ((pos-elem-proj v) neg-party))
+                 (λ (v) (contract-stream-rest v ctc blame))
+                 impersonator-prop:contracted ctc
+                 impersonator-prop:blame (blame-add-missing-party blame neg-party))))))))
+
+(define ((ho-projection impersonate/chaperone-stream) ctc)
+  (let ([elem-ctc (base-stream/c-content ctc)])
+    (λ (blame)
+      (let ([pos-elem-proj ((contract-projection elem-ctc) blame)]
+            [neg-elem-proj ((contract-projection elem-ctc) (blame-swap blame))])
+        (λ (val)
```

Comment:
Use blame-add-context (or one of its variants) here. 

To get a sense of what to put, do something like (list/c (stream/c (box/c none/c))) and look at the error message. You should seem some context messages there. (Let me know if that's not enough of a hint.)